### ISSUE TITLE
zdb routing table

### ISSUE THREAD

**Issue by muhamadazmy** [zero-os/0-core#718]
ZOS must maintain a set of zdb clusters to be used to fetch file blocks a single cluster is defined as

```javascript
{
   <hash-range>: [<endpoint>, ...],
   <hash-range>: [<endpoint>, ...],
}
```
A `<hash-range>` format is (`XX:YY` or `XX`) which matches hash prefix. 
An `<endpoint>` is a data endpoint address for example (zdb://host:port , ardb://host:port)
defaults to zdb protocol if no protocol were given (for example a `hub.gig.tech`)

Example:
```javascript
{
   '00:FF': 'ardb://hub.gig.tech:16379',
}  
``` 
Defaults to current implementation.

Another Example:
```javascript
{
   '00:66': 'zdb://1.endpoint.tf:1234',
   '67:FF': 'zdb://2.endpoint.tf:1234',
}
```

## Usage
Each pool is given a name, and you can define as many pools as you want. How the system uses those pools to look up a `hash` depends on a look up table (ordered list)

```javasctript
table = {
  route: [
    'local',
    'remote'
  ],
 cache: [
    'local'
 ]
}
```

This routing table above, defines a look up order `route`, so a the pool named `local` is checked first, then a `remote` pool

After a block is retrieved, is pushed to pools defined in `cache` section. 

> Take must be taken to avoid blocks to pools that were used to fetch them originally. Also do we need multiple cache pools? or we should limit it to one?

**Comment by muhamadazmy** [comment]
The pools and routing table should be configurable on the fly via the client. Support pushing a full toml/yaml file with the config to set it up in one go.

**Comment by zaibon** [comment]
Practically I don't understand how this type of routine table would work.
If this needs to be set be the client, it means that the routing tables are empty when 0-os boots? That doesn't sounds doable or scalable to me. It seems to be me there should be away for 0-os node to exchange this kind of information so a fresh node could populate its routine table automatically.

Also, when some new data a pushed to the 0-db, how would the routing table of all the node in the network would get this new information ?

Practically what tool would allow to populate the 0-db ? From the format I see that a 0-db is only responsible for a part of a hash (which hash by the way ? ) So we should have some kind of router that when storing data into a list of 0-db, would direct the data to the correct 0-db based on the hash of the data. Which means you already need a routing table to be able to populate the 0-db in the first place. Otherwise you don't know where to send your data to.

**Comment by muhamadazmy** [comment]
Well let me reply for each question one by one.
- ZOS will probably have an initial table that is backward compatible to what we have
Example
```yaml
bools:
  hub:
    - '00:FF':
      - 'ardb://hub.gig.tech:16379'
route:
   - hub
```
Since ZOS must maintain it's `generic` behaviour it should not be tailored to our exact use case. 
- 0-robot when it starts it can then alter the table to add more pools and change it's config as it sees fit. Same way it initialise the storage pools and such.

- New data pushed to 0-db doesn't mean u need to update the routing table, as long that data is pushed correctly. New data though will be cached to local 0-db cluster (if configured).

- Regarding to your question about how to populate the zdb, what I understand 0-db should be a drop-in replacement for ardb (one of the modes have that). In that case, we can use the flist infrastructure we have at the moment (the hub) but upload to the 0-db cluster. A similar routing table in the cluster must define where the blocks should be stored, and then zos can find the blocks once requested.

**Comment by zaibon** [comment]
> 0-robot when it starts it can then alter the table to add more pools and change it's config as it sees fit. Same way it initialise the storage pools and such.

This is still all hard-coded value. So this means there is no automatic way for a 0-db to know about new source of data, right ? Everything needs to be triggered manually?

> New data pushed to 0-db doesn't mean u need to update the routing table, as long that data is pushed correctly. New data though will be cached to local 0-db cluster (if configured).

If you always use a routing table that cover the full hash range (00:FF), this is true, but in any other cause it is not.

>  In that case, we can use the flist infrastructure we have at the moment (the hub) but upload to the 0-db cluster

As far as I understood the idea, we need to get rid of the hub cause it is centralized. So you won't have a single point where to push you data and that will populate 0-db automatically. So that's why I'm asking about the plan regarding the tooling around this. If we want this to work, we also need to work on the design of the tools to make this work.

**Comment by muhamadazmy** [comment]
I suggest we implement this as a first stage, since we need to be able to allow easy configuration or changing of your 'data' source on the fly, next step we can add `dynamic` and auto discovery of other data blocks. 

My only concern about the dynamic configuration, when u try to `know` about all available pools, you can end up with so many pools, which will cause the a huge latency to search all pools `O(n)` to find which one has the block you want. 

> SIDE NOTE: I think a block hash should have more info about the block than the hash. Like a unique namespace so at least, we find out which subset of pools might have this block. 

The problem with hashes that we can't `group` them a single pool can have hashes that span the entire prefix range `00:FF`. Unlike IPs where it's classified. Hence `routing` is not possible the way a router can route certain set of IPs to a gateway that knows more about the IPs final destination.

We can apply the same principle on the hash by adding more data to it (JWT ?) where we have an indicator of the next pool that can have this hash, and if not it will know how to find it. We then can assign  those classified, and dividable ranges to providers. Like what IANA do with IPs.

SIDE NOTE: May be we use the dns names that is (or gonna be) built in digital.me ?!

**Comment by muhamadazmy** [comment]
Okay, so we got this wrong, the idea is to have the config per flist, not system wide. so flists can define their own "routing" to access hashes, so people can ship an flist that embed a config for where to retrieve the hashes. 

But this doesn't explain how we gonna setup the local cache ardb instance ? should we merge the flist config with caching config that defines the local cache ? and add the local cache as a top entity in the lookup table ?

**Comment by zaibon** [comment]
No ardb, we use 0-db. As fas as I know @maxux had made some change so we can use as a source of data for flist as well

**Comment by muhamadazmy** [comment]
@zaibon yeah, for me it's the same interface, I am asking about the config for local cache. if the flist provides the routing tables, how we gonna setup a cache. Does my suggestion make sense ? or you have other ideas?

**Comment by zaibon** [comment]
no I don't understand you question. The cache is not changed, as far as I know 0-fs still need a disk with the files on it to be able to work. So local caching is not impacted here, is it ?
